File: gitlfs-server/src/test/java/ru/bozaro/gitlfs/server/EmbeddedHttpServer.java
package ru.bozaro.gitlfs.server;

import org.eclipse.jetty.server.HttpConnectionFactory;
import org.eclipse.jetty.server.Server;
import org.eclipse.jetty.server.ServerConnector;
import org.eclipse.jetty.servlet.ServletHandler;
import org.eclipse.jetty.servlet.ServletHolder;

import javax.annotation.Nonnull;
import javax.servlet.Servlet;
import java.net.URI;
import java.net.URISyntaxException;

/**
 * Embedded HTTP server for servlet testing.
 *
 * @author Artem V. Navrotskiy
 */
public class EmbeddedHttpServer implements AutoCloseable {
  @Nonnull
  private final Server server;
  @Nonnull
  private final ServerConnector http;
  @Nonnull
  private final ServletHandler servletHandler;

  public EmbeddedHttpServer() throws Exception {
    this.server = new Server();
    this.http = new ServerConnector(server, new HttpConnectionFactory());
    http.setPort(0);
    http.setHost("127.0.1.1");
    server.addConnector(http);

    this.servletHandler = new ServletHandler();
    server.setHandler(servletHandler);

    server.start();
  }

  @Nonnull
  public URI getBase() {
    try {
      return new URI("http", null, http.getHost(), http.getLocalPort(), null, null, null);
    } catch (URISyntaxException e) {
      throw new IllegalStateException(e);
    }
  }

  public void addServlet(@Nonnull String pathSpec, @Nonnull Servlet servlet) {
    servletHandler.addServletWithMapping(new ServletHolder(servlet), pathSpec);
  }

  @Override
  public void close() throws Exception {
    server.stop();
    server.join();
  }
}

File: gitlfs-server/src/test/kotlin/ru/bozaro/gitlfs/server/EmbeddedHttpServer.kt
package ru.bozaro.gitlfs.server

import jakarta.servlet.Servlet
import org.eclipse.jetty.server.HttpConnectionFactory
import org.eclipse.jetty.server.Server
import org.eclipse.jetty.server.ServerConnector
import org.eclipse.jetty.servlet.ServletHandler
import org.eclipse.jetty.servlet.ServletHolder
import java.net.URI
import java.net.URISyntaxException

/**
 * Embedded HTTP server for servlet testing.
 *
 * @author Artem V. Navrotskiy
 */
class EmbeddedHttpServer : AutoCloseable {
    private val server: Server = Server()
    private val http: ServerConnector = ServerConnector(server, HttpConnectionFactory())
    private val servletHandler: ServletHandler

    val base: URI
        get() = try {
            URI("http", null, http.host, http.localPort, null, null, null)
        } catch (e: URISyntaxException) {
            throw IllegalStateException(e)
        }

    fun addServlet(pathSpec: String, servlet: Servlet) {
        servletHandler.addServletWithMapping(ServletHolder(servlet), pathSpec)
    }

    @Throws(Exception::class)
    override fun close() {
        server.stop()
        server.join()
    }

    init {
        http.port = 0
        http.host = "127.0.1.1"
        server.addConnector(http)
        servletHandler = ServletHandler()
        server.handler = servletHandler
        server.start()
    }
}
